src/hpa.c

```diff
@@ -433,16 +457,29 @@ hpa_shard_maybe_do_deferred_work(tsdn_t *tsdn, hpa_shard_t *shard,
 	 * be done.  Otherwise, bound latency to not be *too* bad by doing at
 	 * most a small fixed number of operations.
 	 */
-	size_t maxloops = (forced ? (size_t)-1 : 8);
+	bool hugified = false;
+	bool purged = false;
+	size_t max_ops = (forced ? (size_t)-1 : 16);
+	size_t nops = 0;
 	do {
-		hugified = hpa_try_hugify(tsdn, shard);
-		malloc_mutex_assert_owner(tsdn, &shard->mtx);
+		/*
+		 * Always purge before hugifying, to make sure we get some
+		 * ability to hit our quiesence targets.
+		 */
 		purged = false;
-		if (hpa_should_purge(tsdn, shard)) {
+		while (hpa_should_purge(tsdn, shard)) {
```

Comment:
Don't we lose the max_ops bound here?

Reply:
Added an extra check of max_ops.

In reality, it's a no-op change, since if deferral is allowed then it's the background thread in this loop, and the background thread can continue as long as there's work to be done. If deferral is not allowed, then if we're supposed to purge then it's because the current thread was the one that generated the dirty pages that put us "over the edge", so we're executing a bounded number of times anyways (the dirtiest huge extent is at least as dirty as whatever huge extent we just touched, and so after one purge loop we're back below the threshold).

(This isn't completely true because there can be races in an edge case where another thread deallocates from a huge extent that's mid-purge, putting the hpa_shard over the limit but not triggering a purge itself; in such cases the original thread would have to re-purge it, but this can still just happen a bounded number of times since allocations are banned from a huge extent that's mid-purge).

But in any case, it's still more obviously correct to just have the limit explicit and always-checked.